README.md
beacon-reduc
#############

IRAF tools for polarimetry and spectroscopy from the beacon group

http://beacon.iag.usp.br/


HISTORY
Pacote IRAF Beacon-reduc
############################
Log das modificações das tarefas
==================================

ver 2.1.5, Moser
-----------------
1. Atualizado ``scripts/extract_ms_iraf.py``

2. Movido os scripts não-dependentes de instrumento (como o ``shift_spec.py`` para o ``pyhdust``)

3. Atualizado o *README*, com informações para instalação.

ver 2.1.4, Moser
-----------------
1. Todos os caminhos que apontavam para `/iraf/iraf-v2.16/` foram substituídos por `iraf$`

2. Tudo relacionado com espectroscopia foi atualizado. 

3. Estrutura subpastas: 
  
  a. 'lamps' contém informações sobre as lâmpadas ECASS e MUSICOS
  b.  'scripts' contém executáveis Python e shell

4. Rotina `logpol` renomeada para `pollog`. Isto porque a rotina indepedia digitar `log` para sair do ambiente IRAF. 


ver 2.1.3, Bednarski
---------------------------
1. ccdrap_301.cl: adicionado o mesmo parâmetro 'frame' que fora adicionado já para o ccdrap.cl na versão 2.1.0.


ver 2.1.2, Bednarski
---------------------------
1. ccdrap.cl: adicionado o parâmetro boleano "ver1stwp", que pergunta ao usuário se ele quer verificar se a primeira posição de lâmina está registrada como L0 no cabeçalho.

2. reduce.cl:  

  a. adicionado o mesmo parâmetro "ver1stwp" para setar diretamente esse parâmetro no ccdrap.
  b. agora verifica se o parâmetro reject é MENOR que o limite pré-estabelecido através das configurações dos headers.


ver 2.1.1, Bednarski
---------------------------
1. suape: consertado, com o parâmetro "convert=ascii" retirado, pois estava causando bugs.

2. calib_301.cl: adicionei "add=yes" na chamado do hedit para editar o parâmetro DARKTIME, pois se este campo já não existia, não estava sendo criado automaticamente!

3. ccdrap_301.cl: vários bugs corrigidos.

4. reduce_301.cl: agrupa 8 posições de lâmina também, como a tarefa "reduce.cl".

5. jdrap.cl: swap chamado com tcheck=no para evitar bugs.


ver 2.1.0, Bednarski
---------------------------
1. reduce.cl: verifica se os ganhos (pre-amp) no cabeçalho das séries são iguais aos do cabeçalho das calibrações que serão utilizadas.

2. ccdrap.cl: 

  a. verifica se a primeira posição de lâmina da série é de fato a L0.
  b. acrescentei parâmetro "frame" sobre em qual frame mostrar as imagens no DS9. Isso é útil para reduções em paralelo (com dois terminais cl abertos simultaneamente, usando dois frames diferentes do mesmo DS9).


ver 2.0, Bednarski e Moser
---------------------------
1. ccdrap.cl (Moser e Bednarski)

  a. jun2014: O IRAF 2.16 não aceita mais * como indicador para output de rotinas como referência ao * no parâmetro de entrada. Implementei as modificações do Moser sobre meu ccdrap.cl de substituição dos * por listas de arquivos (@lista).
  b. jun2014: O arquivo que guarda as coordenadas do alvo agora é independente da versão .1, .2, .3, etc. Assim, não é mais preciso mais selecionar os objetos ordinário e extraordinário várias vezes para as versões diferentes.
  c. ago2014: Acrescentei parâmetros para entrada de um arquivo de coordenadas, caso desejado, no qual o daoedit se baseará para encontrar os feixes ordinário/extraordinário (os parâmetros novos são coordref, coord e icom). Esses parâmetros são úteis para o reduce indentificar automaticamente as coordenadas nos outros filtros a partir do primeiro, devendo apenas o usuário confirmá-las. O shell script "icom" é necessário para a entrada do daedit e seu caminho deve ser referenciado no parâmetro homônimo do ccdrap.
  d. ago2014: Adicionei a impressão dos novos parâmetros no ccdraplog (assim como stack1st que não havia sido adicionado).
  e. set2014: Consertei o modo interativo (intera=yes). O mesmo script "icom" do item 1.3 acima busca pela coordenada atualizada da posição de lâmina a partir da posição de lâmina anterior, ao reportar ao usuário se está certo.
  f. mar2015: Comentei umas linhas de impressão de testes que havia esquecido há tempos.
  g. mai2015: Coloquei critérios para calcular se um dos feixes selecionados em uma posição do CCD é perdida ao longa da redução. Isso é feito de duas maneiras: 1) As contagens dos centróides da primeira posição de lâmina são guardadas em uma variável e caso alguma das imagens de uma posição de lâmina dê contagens 10 vezes menor que essa primeira, é porque o algoritmo perdeu o alvo. Veja que isso pode ser repensado em termo das contagens de céu, pois o objeto pode praticamente sumir se houver muitas nuvens; 2) Às vezes um dos feixes é perdido e associado erroneamente a outro feixe e para isso, é verificado se para uma mesma imagem há dois feixes diferentes associados às mesmas coordenadas (observe que se há os feixes ordinário e extraordinário e, hipoteticamente, se invertam em uma imagem, o algoritmo adicionado NÃO vai reportar o ocorrido). Isso é importante porque algumas vezes a redução não é assistida pelo usuário, o que pode deixar passar uma perda nas coordenadas. Ainda não dei um "break" no laço quando um objeto é perdido, mas este é o objetivo após assegurar que a implementação está funcionando. IMPORTANTE: Ir testando nas próximas reduções antes de incorporar de vez no ccdrap.
  h. mai2015: Acrescentei umas mudanças sobre os arquivos de coordenadas caso existentes durante uma nova rodada (pra funcionar um pouco melhor se intera==yes)
  i. jun2015: Agora identifica as coordenadas da n posições de lâmina através das coordenadas da posição imediatamente anterior. Até então era usado como referência apenas as coordenadas da primeira posição quando intera==no. Com isso, espera-se deixar de perder as coordenadas da estrela ao longo das posições de lâmina nos casos de mau acompanhamento.
  j. jun2015: Consertei para excluir arquivos espúrios de rodadas anteriores no início do código.
  k. jun2015: Corrigi um bug: o caminho para o dark e o valor de darksec não estavam sendo guardados em darki e darkseci!
  

2. reduce.cl (Bednarski)

  a.  ago2014: Adicionei ganho e ruído de leitura como parâmetros do reduce.

scripts/norm_spec.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

""" Program to normalize a FITS (*.fit*) or a 2-columns text file (other 
extensions) spectrum based on non-parametric fitting.
"""

from glob import glob
import numpy as np
import os
import sys
import pyfits as pf
import pyhdust.spectools as spt
from argparse import ArgumentParser

__version__ = "0.9"
__author__ = "Daniel Moser"
__email__ = "dmfaes@gmail.com"


class MyParser(ArgumentParser): 
    def error(self, message):
        sys.stderr.write('# ERROR! %s\n' % message)
        self.print_help()
        sys.exit(2)

parser = MyParser(description=__doc__)
parser.add_argument('--version', action='version', 
    version='%(prog)s {0}'.format(__version__))
parser.add_argument("INPUT", help=("String to retrive the spectrum(a) "
    "filename(s) (wildcards accepted)"))
# parser.add_argument("-l", "--line", action="store", dest="line", 
#     help=("Wavelength of the reference line [default: %(default)s]"), 
#     type=float, default=6562.79)

args = parser.parse_args()


def norm_text(fname):
    f0 = np.loadtxt(fname)
    ny = spt.normalize_spec(f0[:, 0], f0[:, 1])
    outname = '{0}_norm{1}'.format(*os.path.splitext(fname))
    f0 = np.savetxt(outname, np.column_stack(f0[:, 0], ny))
    print('# {0} saved!'.format(outname))
    return


def norm_spec(fname):
    imfits = pf.open(fname)
    hdr = imfits[0].header
    flux = imfits[0].data
    wl = np.arange(len(flux)) * hdr['CDELT1'] + hdr['CRVAL1']
    nflux = spt.normalize_spec(wl, flux)
    outname = '{0}_norm{1}'.format(*os.path.splitext(fname))
    hdu = pf.PrimaryHDU(nflux)
    hdulist = pf.HDUList([hdu])
    hdulist[0].header = hdr
    hdu.writeto(outname, clobber=True)
    print('# {0} saved!'.format(outname))
    imfits.close()
    return

if __name__ == '__main__':

    linput = args.INPUT

    for specfile in glob(linput):
        if specfile.find('.fit') > 0:
            norm_spec(specfile)
        else:
            norm_text(specfile)

    if len(glob(linput)) == 0:
        print('# ERROR! {0} files were not found!'.format(linput))


scripts/plot_spec.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

""" Program to export as figure line profiles from standard FITS spectra.
"""

import sys
from glob import glob
import pyhdust.spectools as spt
import pyhdust.phc as phc
from argparse import ArgumentParser
import os
import matplotlib.pyplot as plt

__version__ = "0.9"
__author__ = "Daniel Moser"
__email__ = "dmfaes@gmail.com"


class MyParser(ArgumentParser): 
    def error(self, message):
        sys.stderr.write('# ERROR! %s\n' % message)
        self.print_help()
        sys.exit(2)

parser = MyParser(description=__doc__)
parser.add_argument('--version', action='version', 
    version='%(prog)s {0}'.format(__version__))
parser.add_argument("INPUT", help=("String to retrive the spectrum(a) "
    "filename(s) (wildcards accepted wrapped with \"\")"), type=str)
parser.add_argument("-l", "--line", action="store", dest="line", 
    help=("Wavelength of the reference line [default: %(default)s]"), 
    type=float, default=6562.79)
parser.add_argument("-w", "--hwidth", action="store", dest="hwidth", 
    help=("Half-width of the line profile (in km/s) [default: %(default)s]"), 
    type=float, default=1000.)

# group2 = parser.add_argument_group('other arguments (overwrite power)')
# group2.add_argument("-r", "--remove", action="store_true", dest="rm", 
#     help=("If this flag is enabled, it restores the spectrum(a) to the "
#         "original value (i.e., VELSHIFT = 0)"), default=False)
# group2.add_argument("-f", "--force-dv", action="store", dest="vs", 
#     help=("Force VELSHIFT to the non-zero VS value (in km/s) "
#         "[default: None]"), type=float, default=0.)

args = parser.parse_args()


if __name__ == '__main__':

    lbc = args.line
    linput = args.INPUT
    hwidth = args.hwidth

    plev = 0.
    lfiles = glob(linput)
    lfiles.sort()
    if len(lfiles):
        fig, ax = plt.subplots(figsize=(8, 6*len(lfiles)/25))

    for fitsfile in lfiles:
        sdata = spt.loadfits(fitsfile)
        vl, fx = spt.lineProf(sdata[0], sdata[1], lbc=lbc, hwidth=hwidth)
        ax.plot(vl, fx+plev)
        ax.text(0.95*hwidth, 1.+plev+0.05/2, os.path.splitext(fitsfile)[0], 
            horizontalalignment='right',  # transform=ax.transAxes,
        verticalalignment='center', fontsize=8)
        if plev == 0:
            ymin = min(fx)
        plev += 0.04

    if len(glob(linput)) == 0:
        print('# ERROR! {0} files were not found!'.format(linput))
    else:
        # ax.legend(loc='best', fancybox=True, framealpha=0.5, fontsize=8, 
        #         labelspacing=0.05)
        ax.set_ylim([ymin, plev+max(fx)])
        ax.set_xlim([-hwidth, hwidth])
        phc.savefig(fig, figname='plot_spec_'+phc.dtflag())


scripts/shift_spec.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

""" Program that shifts a FITS spectrum based on an Halpha Gaussian fit.
"""

from glob import glob
import numpy as np
import sys
import pyfits as pf
import pyhdust.spectools as spt
import pyhdust.phc as phc
from scipy.optimize import curve_fit
from argparse import ArgumentParser

__version__ = "0.9"
__author__ = "Daniel Moser"
__email__ = "dmfaes@gmail.com"


class MyParser(ArgumentParser): 
    def error(self, message):
        sys.stderr.write('# Error: %s\n' % message)
        self.print_help()
        sys.exit(2)


parser = MyParser(description=__doc__)
parser.add_argument('--version', action='version', 
    version='%(prog)s {0}'.format(__version__))
parser.add_argument("INPUT", help=("String to retrive the spectrum(a) "
    "filename(s) (wildcards accepted)"))
parser.add_argument("-l", "--line", action="store", dest="line", 
    help=("Wavelength of the reference line [default: %(default)s]"), 
    type=float, default=6562.79)

args = parser.parse_args()


def gauss(x, a, x0, sigma, y0=1.):
    return a*np.exp(-(x-x0)**2/(2*sigma**2))+y0


def find_max_peak(vl, nflx, idx=-1):
    """ Based on max flux. 
    """
    if idx == -1:
        idx = np.where(nflx == np.max(nflx))
        return nflx[idx], vl[idx]
    else:
        idx_ = np.where(nflx[:idx] == np.max(nflx[:idx]))
        nfV, vlV = nflx[idx_], vl[idx_]
        idx_ = np.where(nflx[idx:] == np.max(nflx[idx:]))
        nfR, vlR = nflx[idx_], vl[idx_]
    return nfV, vlV, nfR, vlR


def has_central_absorp(nf0, vlV, vlR):
    """ Return ``True`` or ``False`` if the line has a central absorption 
    (i.e., it is an absorption line or it is an double peak emission line).
    """
    if nf0 < 1:
        # print('# F(lb=0) < 1: A double-peak emission line was found!')
        return True
    if vlR-vlV > 100:
        return True
    else:
        return False


def gauss_fit(x, y, a0=None, x0=None, sig0=None, y0=None, emission=True):
    """ Return ``curve_fit``, i.e., ``popt, pcov``.
    """
    if a0 is None:
        a0 = np.percentile(y, 95) - np.mean(y)
        if not emission:
            a0 *= -1
    if x0 is None:
        x0 = np.sum(x*y)/len(x)
    if sig0 is None:
        sig0 = (x[-1]-x[0])/10.
    if y0 is None:
        q = 5
        if not emission:
            q = 95
        y0 = np.percentile(y, q)
    return curve_fit(gauss, x, y, p0=[a0, x0, sig0, y0])

lbc = args.line
linput = args.INPUT

for fitsfile in glob(linput):

    imfits = pf.open(fitsfile, mode='update')
    flux = imfits[0].data
    wl = np.arange(len(flux)) * imfits[0].header['CDELT1'] + \
        imfits[0].header['CRVAL1']

    vl, nflx = spt.lineProf(wl, flux, lbc=lbc)
    idx0 = phc.find_nearest(vl, 0., idx=True)
    nf0 = nflx[idx0]
    nfV, vlV, nfR, vlR = find_max_peak(vl, nflx, idx=idx0)

    if has_central_absorp(nf0, vlV, vlR):
        idxV = phc.find_nearest(vl, vlV, idx=True)
        idxR = phc.find_nearest(vl, vlR, idx=True)+idx0
        popt, pcov = gauss_fit(vl[idxV:idxR], nflx[idxV:idxR], emission=False)
    else:
        popt, pcov = gauss_fit(vl, nflx)

    vshift = -popt[1]
    new_wl = wl*(1 + vshift/phc.c.cgs*1e5)

    imfits[0].header['CDELT1'] = (new_wl[-1]-new_wl[0])/(len(new_wl)-1)
    imfits[0].header['CRVAL1'] = new_wl[0]
    # imfits[0].data = np.interp(new_wl, wl, flux)

    old_sh = 0
    if 'VELSHIFT' in imfits[0].header:
        old_sh = imfits[0].header['VELSHIFT'] 
    imfits[0].header['VELSHIFT'] = (vshift + old_sh, 'Lambda shift in km/s')

    print('# {0} updated with {1:.1f} km/s shift (accum. {2:.0f} km/s)'.format(
        fitsfile, vshift, vshift+old_sh))
    imfits.close()

if len(glob(linput)) == 0:
    print('# ERROR! {0} files were not found!'.format(linput))


scripts/grafpol.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

"""
Program to plot the best adjust and its residuals.
pyhdust package is required.

grafpol [path] [suf] [extension]

'path' is path to the log files. If not specified, it is suposed as '.' dir.
'suf' is to generate only the graphs for log files having 'suf' suffix. 
'suf' must begin with one '_' character. If not specified, generate for all 
logfile found.
'extension' is the extension for the graphs ('png', 'eps' or 'jpg'). If not 
specified, is used 'png'.


Daniel Bednarski, June 2015
"""

import os
import numpy as np
from glob import glob
from sys import argv

try:
    import pyhdust.poltools as polt
except ImportError:
    print('# ERROR: pyhdust package not installed!')
    raise SystemExit(1)

# default values
path = os.getcwd()
extens = 'png'
suf = ''

# argv[0] is the command name
for i in range(1, len(argv)):
    if argv[i][0] == '_':
        suf = argv[i]
    elif argv[i] in ('eps', 'png', 'jpg'):
        extens = argv[i]
    else:
        path = argv[i]
        if not os.path.exists(path):
            print('# ERROR: path {0} doesn\'t exist!'.format(path))
            raise SystemExit(1)

logs = glob('{0}/*{1}_*.log'.format(path, suf))
if len(logs) == 0:
    print('# ERROR: no *.log files found!')
    print('{0}/*{1}_*.log'.format(path, suf))
    raise SystemExit(1)

print('# Please wait, generating graphs...')
for log in logs:
    f0 = np.loadtxt(log, dtype=str, delimiter='\n', comments=None)
    nstars = int(f0[6].split()[-1])
    for nstar in range(1, nstars + 1):
        polt.grafpol(log, nstar=nstar, save=True, extens=extens)

print('# Done!')


scripts/create_ut.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

""" Program to add the *UT* entry the in fits header. It requires the 
*DATE-OBS* entry. 
"""
import pyfits as pf
from glob import glob
import numpy as np
import jdcal

__version__ = "0.9"
__author__ = "Daniel Moser"
__email__ = "dmfaes@gmail.com"


def hms2fracday(hms):
    """ Enter hour:min:sec (string) and return fraction of a day (float) """
    hms = np.array(hms.split(':'), dtype='float')
    return (hms[0] + hms[1]/60. + hms[2]/3600.) / 24.


def longdate2MJD(ldate):
    """ FROM YYYY-MM-HHThh:mm:ss.sss to MJD (float). """
    ldate, hms = ldate.split('T')
    ldate = np.array(ldate.split('-'), dtype=int)
    mjd = jdcal.gcal2jd(*ldate)[1]
    return mjd + hms2fracday(hms)


if __name__ == '__main__':
    lfits = glob('*.fit*')
    for fn in lfits:
        f = pf.open(fn, mode='update')
        if 'DATE-OBS' in f[0].header and 'UT' not in f[0].header:
            ldate, hms = f[0].header['DATE-OBS'].split('T')
            f[0].header['UT'] = hms
            f.flush()
            print('# {0} updated with UT!'.format(fn))
        elif 'DATE-OBS' not in f[0].header:
            print('# ERROR! DATE-OBS not in header of {0}'.format(fn))
            if 'DATE' in f[0].header and 'UT' not in f[0].header:
                ldate, hms = f[0].header['DATE'].split('T')
                f[0].header['UT'] = hms
                f.flush()
                print('# {0} updated with UT!'.format(fn))                
        else:
            print('# {0} already contains UT info.'.format(fn))
        f.close()


scripts/renfilters.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

""" Program to rename the IAGPOL filters id from "_Fn" to "_f", where f is one
of the filters (u, b, v, r, i).
"""
from glob import glob
import os

__version__ = "0.9"
__author__ = "Daniel Moser"
__email__ = "dmfaes@gmail.com"


def renlist(root, newr):
    """ The routine changes each A_STR_B to A_NEW_B inside the running folder.
    """
    files = glob('*{0}*'.format(root))
    files.sort()
    for i in range(len(files)):
        os.system(
            'mv "' + files[i] + '" "' + files[i].replace(root, newr) + '"' )
        print("# " + files[i] + " renamed to: " +
              files[i].replace(root, newr) )
    return


if __name__ == '__main__':
    print("# This will change the filters F# to [u,b,v,r,i] letters.")
    roots = ['_F0', '_F1', '_F2', '_F3', '_F4']
    newrs = ['_u', '_b', '_v', '_r', '_i']

    for i in range(len(roots)):
        renlist(roots[i], newrs[i])


scripts/extract_ms_iraf.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-

""" Program to normalize spectra from IRAF Echelle format (*.ms.cal.fits) based 
on spectrum derivative variations.
"""
import os
from glob import glob
import numpy as np
import sys
# import pyhdust.spectools as spt
# import pyhdust.phc as phc
from scipy import interpolate
import matplotlib.pyplot as plt
from scipy.signal import savgol_filter
import specutils.io as sio
import pyfits
import datetime as dt
from argparse import ArgumentParser

__version__ = "0.9"
__author__ = "Daniel Moser"
__email__ = "dmfaes@gmail.com"


class MyParser(ArgumentParser): 
    def error(self, message):
        sys.stderr.write('# Error: %s\n' % message)
        self.print_help()
        sys.exit(2)

parser = MyParser(description=__doc__)
parser.add_argument('--version', action='version', 
    version='%(prog)s {0}'.format(__version__))
# parser.add_argument("INPUT", help=("String to retrive the spectrum(a) "
#     "filename(s) (wildcards accepted)"))
# parser.add_argument("-l", "--line", action="store", dest="line", 
#     help=("Wavelength of the reference line [default: %(default)s]"), 
#     type=float, default=6562.79)
parser.add_argument("-F", "--fig-fmt", action="store", dest="fig_fmt", 
    help=("Format of the output figures [default: %(default)s]"), 
    type=str, default="png")

args = parser.parse_args()


def dtflag():
    """ Return a "datetime" flag, i.e., a string the the current date and time
    formated as yyyymmdd-hhMM."""
    now = dt.datetime.now()
    return '{0}{1:02d}{2:02d}-{3:02d}{4:02d}{5:02d}'.format(now.year, 
        now.month, now.day, now.hour, now.minute, now.second)


def sum_ec(fwl, fflx):
    dmin = np.inf
    wlmin = np.inf
    wlmax = 0
    for f in fwl:
        if np.min(np.diff(f)) < dmin:
            dmin = np.min(np.diff(f))
        if np.min(f) < wlmin:
            wlmin = np.min(f)
        if np.max(f) > wlmax:
            wlmax = np.max(f)
    swl = np.arange(wlmin, wlmax, dmin)
    sflx = np.zeros(len(swl))
    for i in range(len(fwl)):
        idx = np.where( (swl > np.min(fwl[i])) & (swl < np.max(fwl[i])) )
        sflx[idx] += np.interp(swl[idx], fwl[i], fflx[i])
    return swl, sflx


def plot_orders(ax, fwl, fflx, mode='o'):
    for i in range(len(fwl)):
        ax.plot(fwl[i], fflx[i], mode, ms=1)
    return ax


def splitequal(n, N):
    """ Split `N` in approx. `n` igual parts 

    `N` must be integer.

    Suggestion: *phc.splitequal(N/8., N)* split N into sequences of 
    approx. 8 itens.
    """
    n = int(round(n))
    idx = []
    for i in range(n):
        idx.append([i * N / n, (i + 1) * N / n])
    if n == 0:
        idx = [[0, N]]
    return idx


def max_func_pts(x, y, ws=0.01, avgbox=3):
    """ `ws` window size where the maximum will be evaluated. Example: `ws=0.02`
    corresponds to 2% of the length of the input. """
    x, y = (np.array(x), np.array(y))
    N = len(x)
    parts = splitequal(N*ws, N)
    n = len(parts)
    xout, yout = (np.zeros(n), np.zeros(n))
    for i in range(n):
        p = parts[i]
        Y = y[p[0]:p[1]]
        X = x[p[0]:p[1]]
        idx = np.argsort(Y)
        xout[i] = np.average(X[idx][-avgbox:])
        yout[i] = np.average(Y[idx][-avgbox:])
    return xout, yout


def do_norm(wl, flx):
    xout, yout = max_func_pts(wl, flx, ws=0.01, avgbox=5)
    tck = interpolate.splrep(xout, yout)
    ynew = interpolate.splev(xout, tck, der=1)
    dmed = np.median( np.abs(ynew) )
    l = len(wl)
    # ConditionA: small derivative 
    conda = ( np.abs(ynew) < dmed )  
    # ConditionB: in the limits
    condb = ( wl[int(l*0.15)] > xout ) | ( xout > wl[int(l*(1-.15))] )
    # xmax = phc.find_nearest(yout, np.max(yout), idx=True)
    # xmax = xout[xmax]
    # xmax = wl[l/2]
    # ConditionC: xout < max(flx) AND ynew > 0
    # condc = ( xout <= xmax ) & ( ynew >= 0 )
    # ConditionD: xout > max(flx) AND ynew < 0
    # condd = ( xout >= xmax ) & ( ynew <= 0 )
    conde = ( ( xout > 6584.5 ) | ( xout < 6541. ) )
    # idx = np.where( (( conda & (condd | condc) ) | condb) & (conde) )  
    idx = np.where( (conda | condb) & (conde) )  
    # & (condd | condc))
    xout = xout[idx]
    yout = yout[idx]
    tck = interpolate.splrep(xout, yout, k=3)
    wl = wl[2:-2]
    flx = flx[2:-2]
    ynew = interpolate.splev(wl, tck, der=0)
    return wl, flx, ynew


def writeFits(flx, lbd, extrahead=None, savename=None, quiet=False, path=None,
    lbdc=None, externhd=None):
    """ Write a 1D spectra FITS.

    | INPUT: flux array, lbd array, extrahead flag+info, save name.
    | - lbd array: if len(lbd)==2: lbd = [CRVAL1, CDELT1]
    |              else: CDELT1 = (lbd[-1]-lbd[0])/(len(lbd)-1)
    |                    CRVAL1 = lbd[0]
    |   WARNING: lbd must be in ANGSTROMS (FITS default). It can also be 
    |   velocities. In this case, it must be in km/s and lbdc is given in 
    | ANGSTROM.
    | - extrahead: matrix (n,2). Example: [['OBJECT','Achernar'], ['COMMENT',
    | 'value']]

    `externhd` = copy the header from an external file.

    OUTPUT: write FITS file.
    """
    if path is None or path == '':
        path = os.getcwd()
    if path[-1] != ['/']:
        path += '/'
    if lbdc is not None:
        lbd = (lbd / 2.99792458e10 * 1e5 + 1) * lbdc
    hdu = pyfits.PrimaryHDU(flx)
    hdulist = pyfits.HDUList([hdu])
    if externhd is not None:
        extf = pyfits.open(externhd)
        hdulist[0].header = extf[0].header
        hdulist[0].header['BZERO'] = 0.
    hdulist[0].header['CRVAL1'] = lbd[0]
    if len(lbd) == 2:
        hdulist[0].header['CDELT1'] = lbd[1]
    else:
        hdulist[0].header['CDELT1'] = (lbd[-1] - lbd[0]) / (len(lbd) - 1)
    if extrahead is not None:
        for e in extrahead:
            hdulist[0].header[e[0]] = e[1]
    if savename is None:
        savename = 'spec_{0}'.format(dtflag())
    if savename.find('.fit') == -1:
        savename += '.fits'
    hdu.writeto(path + savename, clobber=True)
    if not quiet:
        print('# FITS file {0}{1} saved!'.format(path, savename))
    return


if __name__ == '__main__':
    path = '.'
    opsps = glob('*.ms.cal.fits')
    print opsps

    # avgspecs = []
    for sp in opsps:
        # os.system('gunzip -c {0} > tmp.txt'.format(sp))
        # spec = np.loadtxt('tmp.txt', skiprows=11)
        spec = sio.read_fits.read_fits_spectrum1d(sp)

        # fig, ax = plt.subplots()
        # ax.plot(spec[:, 5], spec[:, 10], label='Leg.')
        # # ax.set_title('Title')
        # # ax.legend()
        # phc.savefig(fig, figname=sp.replace('.spc.gz', '_ori'))

        # i = 0
        onorm = []
        fwl = []
        oflx = []
        ocont = []
        # orders = np.unique(spec[:, 0])
        for i in range(len(spec)):
            wl = np.array(spec[i].wavelength)
            flx = np.array(spec[i].flux)
            try:
                wl, flx, ynew = do_norm(wl, flx)
                fwl.append(wl)
                onorm.append(flx/ynew)
                ocont.append(ynew)
                oflx.append(flx)        
            except:
                print('# Error in order {0}'.format(i))

        fig, ax = plt.subplots()
        ax = plot_orders(ax, fwl, oflx, 'o')
        ax = plot_orders(ax, fwl, ocont, '-')
        plt.savefig(sp.replace('.fits', '.orders.{0}'.format(args.fig_fmt)))

        swl, sflx = sum_ec(fwl, oflx)
        swl, scont = sum_ec(fwl, ocont)
        swl, snorm = sum_ec(fwl, onorm)

        # avgspecs.append(sp.replace('.spc.gz', '.rv.fits'))
        nans, x = (np.isnan(scont), lambda z: z.nonzero()[0])
        sflx[nans] = 1.
        scont[nans] = 1.
        idx = np.where(scont <= 1e-3)
        scont[idx] = 1.
        sflx[idx] = 1.
        final = sflx/scont
        idx = np.where((final > 0.986) & (final < 1.014))
        pflx = final[idx]
        pwl = swl[idx]
        pflx = np.interp(swl, pwl, savgol_filter(pflx, 3, 1, mode='nearest'))
        final = final/pflx
        writeFits( final, swl, savename=sp.replace('.fits', '.ext.fits'), 
            quiet=False, externhd=sp )

        fig, ax = plt.subplots()
        ax.plot(swl, final)
        ax.set_ylim([0, 3])
        plt.savefig(sp.replace('.fits', '.{0}'.format(args.fig_fmt)))

        ax.set_xlim([6540, 6585])
        ax.set_ylim([0.5, 2.0])
        plt.savefig(sp.replace('.fits', 'halpha.{0}'.format(args.fig_fmt)))
        plt.close(fig)

    # if os.path.exists('tmp.txt'):
    #     os.system('rm tmp.txt')

    # spt.averagespecs(avgspecs)
